src/mlpack/core/cv/metrics/r2_score.hpp

```diff
@@ -43,7 +43,10 @@ namespace cv {
  * For example, a model having R2Score = 0.85, explains 85 \% variability of
  * the response data around its mean.
  */
-class R2Score
+
+template<bool adjustedR2> class R2Score;
+
+template<> class R2Score<false>
```

Comment:
Hey @shawnbrar, thanks for taking the time to add this template parameter!  Actually, I think you don't need template specialization here.  All you should need to do is declare the class as:

```
template<bool AdjustedR2>
class R2Score
```

and then in the implementation of `Evaluate()`, you can change the bottom to this:

```
    if (AdjustedR2)
    {
      // Handling undefined R2 Score when both denominator and numerator is 0.0.
      if (residualSumSquared == 0.0)
        return totalSumSquared ? 1.0 : DBL_MIN;
      // Returning adjusted R-squared.
      double rsq = 1 - (residualSumSquared / totalSumSquared);
      return (1 - ((1 - rsq) * ((data.n_cols - 1) / (data.n_cols - data.n_rows - 1))));
    }
    else
    {
      // Returning R-squared
      return 1 - residualSumSquared / totalSumSquared;
    }
```

and that should be all that's necessary.  The nice thing about templates is that that code above will actually be compiled into two different functions *at compile time*, so the `if (AdjustedR2)` won't actually be run when the program is executed---only the correct branch will be run!

Would you mind refactoring it to try this?  It should result in a significantly shorter diff. :+1:

Reply:
Dear @rcurtin , sure, even I was thinking of a way which would have been shorter but like I said I am not an experienced programmer in C++.

Reply:
> Hey @shawnbrar, thanks for taking the time to add this template parameter! Actually, I think you don't need template specialization here. All you should need to do is declare the class as:
> 
> ```
> template<bool AdjustedR2>
> class R2Score
> ```
> 
> and then in the implementation of `Evaluate()`, you can change the bottom to this:
> 
> ```
>     if (AdjustedR2)
>     {
>       // Handling undefined R2 Score when both denominator and numerator is 0.0.
>       if (residualSumSquared == 0.0)
>         return totalSumSquared ? 1.0 : DBL_MIN;
>       // Returning adjusted R-squared.
>       double rsq = 1 - (residualSumSquared / totalSumSquared);
>       return (1 - ((1 - rsq) * ((data.n_cols - 1) / (data.n_cols - data.n_rows - 1))));
>     }
>     else
>     {
>       // Returning R-squared
>       return 1 - residualSumSquared / totalSumSquared;
>     }
> ```
> 
> and that should be all that's necessary. The nice thing about templates is that that code above will actually be compiled into two different functions _at compile time_, so the `if (AdjustedR2)` won't actually be run when the program is executed---only the correct branch will be run!
> 
> Would you mind refactoring it to try this? It should result in a significantly shorter diff.

Dear @rcurtin , I have removed the template specialization and made it the way you had asked for.